packages/core/contracts/financial-templates/contract-for-difference/ContractForDifference.sol

```diff
@@ -0,0 +1,183 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/utils/math/SafeMath.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/utils/Address.sol";
+
+import "../common/financial-product-libraries/ContractForDifferenceFinancialProductLibrary.sol";
+
+import "../../common/implementation/Testable.sol";
+import "../../common/implementation/Lockable.sol";
+import "../../common/implementation/FixedPoint.sol";
+import "../../common/interfaces/ExpandedIERC20.sol";
+import "../../common/interfaces/IERC20Standard.sol";
+
+import "../../oracle/interfaces/OracleInterface.sol";
+
+import "../../oracle/interfaces/FinderInterface.sol";
+import "../../oracle/interfaces/OptimisticOracleInterface.sol";
+import "../../oracle/interfaces/IdentifierWhitelistInterface.sol";
+
+import "../../oracle/implementation/Constants.sol";
+
+abstract contract ContractForDifference is Testable, Lockable {
+    using FixedPoint for FixedPoint.Unsigned;
+    using SafeMath for uint256;
+    using SafeERC20 for IERC20;
+    using SafeERC20 for ExpandedIERC20;
+    using Address for address;
+
+    enum ContractState { Open, ExpiredPriceRequested, ExpiredPriceReceived }
+    ContractState public contractState;
+
+    uint32 public expirationTimestamp;
+
+    uint256 public collateralPerPair;
+    uint256 public totalCollateral;
+
+    int256 public expiryPrice;
+
+    uint256 public expirationTokensForCollateral;
+
+    bytes32 public priceIdentifier;
+
+    IERC20 public collateralToken;
+    ExpandedIERC20 public longToken;
+    ExpandedIERC20 public shortToken;
+
+    FinderInterface public finder;
+
+    ContractForDifferenceFinancialProductLibrary public financialProductLibrary;
+
+    modifier preExpiration() {
+        require(getCurrentTime() < expirationTimestamp, "Only callable pre-expiry");
+        _;
+    }
+
+    modifier postExpiration() {
+        require(getCurrentTime() >= expirationTimestamp, "Only callable post-expiry");
+        _;
+    }
+
+    modifier onlyOpenState() {
+        require(contractState == ContractState.Open, "Contract state is not OPEN");
+        _;
+    }
+
+    constructor(
+        uint32 _expirationTimestamp,
+        uint256 _collateralPerPair,
+        address _longTokenAddress,
+        address _shortTokenAddress,
```

Comment:
nit: can we just make these parameter types `ExpandedIERC20` for simplicity?

Same suggestion for the other address params here.